proposals/namespace_modules/20-xxx.txt

```diff
@@ -0,0 +1,176 @@
+
+    from A import foo                  use A, only: foo
+    from A import foo as Afoo          use A, only: Afoo => foo
+    from A import *                    use A
+
+Except:
+
+    Python                             Fortran
+
+    import A                           N/A
+    import A as B                      N/A
+
+This proposal proposes to fill in the missing functionality as
+follows:
+
+    Python                             Fortran
+
+    import A                           use, with :: A
+    import A as B                      use, with :: B => A
+
+or perhaps as
+
+
+    Python                             Fortran
+
+    import A                           with A
+    import A as B                      with B => A
+
+
+3. Use Cases
+
+3.1 Same function names in multiple modules
+
+In Python a very common idiom is:
+
+    import math
+    import numpy as np
+    import sympy as sym
+    ...
+    e1 = np.sin(np.pi)      # NumPy expression
+    e2 = math.sin(math.pi)  # Built-in Python math expression
+    e3 = sym.sin(sym.pi)    # SymPy expression
+
+In Fortran currently one has to do:
+
+    use math, only: math_sin => sin, math_pi => pi
+    use numpy, only: np_sin => sin, np_pi => pi
+    use sympy, only: sym_sin => sin, sym_pi => pi
+    ...
+    e1 = np_sin(np_pi)      ! NumPy expression
+    e2 = math_sin(math_pi)  ! Built-in Python math expression
+    e3 = sym_sin(sym_pi)    ! SymPy expression
+
+With this proposal one could also do:
+
+    use, with :: math
+    use, with :: np => numpy
+    use, with :: sym => sympy
+    ...
+    e1 = np%sin(np%pi)      ! NumPy expression
+    e2 = math%sin(math%pi)  ! Built-in Python math expression
+    e3 = sym%sin(sym%pi)    ! SymPy expression
+
+
+3.2 Need to import lots of functions from a module
+
+Existing code (https://github.com/certik/fortran-utils/blob/
+b43bd24cd421509a5bc6d3b9c3eeae8ce856ed88/src/linalg.f90):
+
+    use lapack, only: dsyevd, dsygvd, ilaenv, zgetri, zgetrf, &
+        zheevd,  dgeev, zgeev, zhegvd, dgesv, zgesv, dgetrf, &
+        dgetri, dgelsy, zgelsy, dgesvd, zgesvd, dgeqrf, dorgqr, &
+	dpotrf, dtrtrs
+    ...
+    call dgeev('N', 'V', n, At, lda, wr, wi, vl, ldvl, vr, ldvr, &
+     work, lwork, info)
+    ...
+    call dgetrf(n, n, Amt, lda, ipiv, info)
+    ...
+
+Instead, one can write it as:
+
+    use, with :: lapack
+    ...
+    call lapack%dgeev('N', 'V', n, At, lda, wr, wi, vl, ldvl, vr, &
+     ldvr, work, lwork, info)
+    ...
+    call lapack%dgetrf(n, n, Amt, lda, ipiv, info)
+    ...
+
+Then when another subroutine must be called from the `lapack` module,
+one can just call it, without having to modify the `use` statement.
+
+3.3 Conflicting derived types
+
+    use math, only: math_vector => vector
+    use strings, only: string_vector => vector
+
+    type(math_vector) :: avector
+    type(string_vector) :: bvector
+    ...
+
+vs.
+
+    use, with :: math, only: vector
+    use, with:: strings, only: vector
```

Comment:
I believe this should be just:
```suggestion
    use, with :: math
    use, with :: strings
```
the code below would still run. I don't think we should allow `user, with :: math, only: vector`, what would it do? Allow `math%vector`, but not `math%matrix`? Python doesn't allow that, I wouldn't allow that either here.

Reply:
FWIW I tried going to the file accept the change and nothing seemed to happen. I wonder if Milan has ownership.